pg-extend/src/pg_datum.rs

```diff
@@ -110,6 +110,43 @@ impl From<i64> for PgDatum {
     }
 }
 
+impl TryFromPgDatum for String {
+    fn try_from(datum: PgDatum) -> Result<Self, &'static str> {
+        use std::ffi::CStr;
+        use std::os::raw::c_char;
+
+        if let Some(datum) = datum.0 {
+            let text_val = datum as *const pg_sys::text;
+
+            let cstr = unsafe {
+                let val: *mut c_char = pg_sys::text_to_cstring(text_val);
+                CStr::from_ptr(val)
```

Comment:
I know you said that you have tested and don't think there's a memory leak or anything here, that might be because all memory created with pg_alloc is freed at the end of the transaction/connection. But, can you verify that this cstr doesn't need to be freed? This will matter in the context of any extensions that are working across connections.

In other words, I think text_val might need to be freed after the CStr is converted into a String.

Reply:
Do you mean like this?

```suggestion

            let cstr = unsafe {
                let val: *mut c_char = pg_sys::text_to_cstring(text_val);
                CStr::from_ptr(val)
            };

            let value = match cstr.to_str() {
                Ok(s) => Ok(s.to_owned()),
                Err(_) => Err("datum was not valid utf8")
            };

            drop(cstr);
            value
```

Reply:
Ah, sorry for not being clearer. I'm wondering if `pg_sys::text_to_cstring` allocates a new string that needs to be deallocated.

If it doesn, then I think rather than `CStr::from_ptr(val)`, what we want is `CString::from_raw(val)`. Though, I just read the docs on this: https://doc.rust-lang.org/std/ffi/struct.CString.html#method.from_raw , and it quite clearly states that the it should not be used for taking ownership of the memory.

What this probably means, is that where you have that `drop(cstr)`, we actually need a pfree (I need to find the correct pg function for this) of `val`.

Reply:
Yeah ok, that makes more sense. I've updated it to free the `text_val` pointer.